compose/service.py

```diff
@@ -395,27 +395,57 @@ def _containers_have_diverged(self, containers):
     def execute_convergence_plan(self,
                                  plan,
                                  do_build=True,
-                                 timeout=DEFAULT_TIMEOUT):
+                                 timeout=DEFAULT_TIMEOUT,
+                                 detached=False):
         (action, containers) = plan
 
         if action == 'create':
             container = self.create_container(do_build=do_build)
+
+            if container.has_api_logs and not detached:
+                container.attach_log_stream()
+
             self.start_container(container)
 
             return [container]
 
         elif action == 'recreate':
-            return [
-                self.recreate_container(
-                    c,
-                    timeout=timeout
+            # de-couple recreate_container by copying it here for this use
+            # case to work with ability to attach log_stream *before* we
+            # create a container
+            recreated_containers = []
+            for container in containers:
+                log.info("Recreating %s" % container.name)
+                # stop
+                container.stop(timeout=timeout)
+                container.rename_to_tmp_name()
+                # create
+                new_container = self.create_container(
+                    do_build=False,
+                    previous_container=container,
+                    number=container.labels.get(LABEL_CONTAINER_NUMBER),
+                    quiet=True,
                 )
-                for c in containers
-            ]
+
+                # attach
+                if new_container.has_api_logs and not detached:
+                    new_container.attach_log_stream()
+                recreated_containers.append(new_container)
+
+                # restart
+                self.start_container(new_container)
+                container.remove()
+
+            return recreated_containers
 
         elif action == 'start':
-            for c in containers:
-                self.start_container_if_stopped(c)
+            for container in containers:
+                if not container.is_running:
+                    if container.has_api_logs and not detached:
+                        container.attach_log_stream()
+
+                    log.info("Starting %s" % container.name)
+                    self.start_container(container)
```

Comment:
If we add a flag `attach_logs` to `start_container_if_stopped()` which calls this. I think this would be:

```python

for container in containers:
    self.start_container_if_stopped(container, attach_logs=not detached)
```

which removes the need to duplicate anything